### PR TITLE
Filter DataGrid rows by multiple conditions

### PR DESCRIPTION
**Pull request by c-bata**
## Contributor License Agreement

This repository (``optuna-dashboard``) and [Goptuna](https://github.com/c-bata/goptuna) share common code.
This pull request may therefore be ported to Goptuna.
Make sure that you understand the consequences concerning licenses and check the box below if you accept the term before creating this pull request.

* [x] I agree this patch may be ported to [Goptuna](https://github.com/c-bata/goptuna) by other Goptuna contributors.

## Reference Issues/PRs

Refs #696

## What does this implement/fix? Explain your changes.

Filter DataGrid rows by multiple conditions.

Screenshot 2023-11-16 10 41 51

### CODE REVIEW

**Comment by c-bata** [comment]
I'll make this PR ready for review once #698 will be merged.

**Comment by c-bata** [comment]
@nabenabe0928 Could you review this PR?

File: optuna_dashboard/ts/components/DataGrid.tsx [hunk 1]
```diff
@@ -239,30 +190,119 @@ function DataGrid<T>(props: {
...
+              open={filterMenuAnchorEl !== null}
+              onClose={() => {
+                setFilterMenuAnchorEl(null)
+              }}
+            >
+              {filterChoices.map((choice) => (
+                <MenuItem
+                  key={choice}
+                  onClick={() => {
+                    const values =
```
Comment (nabenabe0928) [inline optuna_dashboard/ts/components/DataGrid.tsx:269]:
```suggestion
                    const newTickedValues =
```

File: optuna_dashboard/ts/components/DataGrid.tsx [hunk 2]
```diff
@@ -239,30 +190,119 @@ function DataGrid<T>(props: {
...
+                      filter === null
+                        ? filterChoices.filter((v) => v !== choice)
```
Comment (nabenabe0928) [inline optuna_dashboard/ts/components/DataGrid.tsx:271]:
```suggestion
                        ? filterChoices.filter((v) => v !== choice)  // By default, every choice is ticked, so the chosen option will be unticked.
```

File: optuna_dashboard/ts/components/DataGrid.tsx [hunk 3]
```diff
@@ -239,30 +190,119 @@ function DataGrid<T>(props: {
...
+                        : filter.values.some((v) => v === choice)
+                        ? filter.values.filter((v) => v !== choice)
+                        : [...filter.values, choice]
+                    onFilterChange(values)
+                  }}
+                >
+                  <ListItemIcon>
+                    {!filter || !filter.values.every((v) => v !== choice) ? (
```
Comment (nabenabe0928) [inline optuna_dashboard/ts/components/DataGrid.tsx:279]: It is more consistent with the code above.
```suggestion
                    {!filter || filter.values.some((v) => v === choice) ? (
```

File: optuna_dashboard/ts/components/DataGrid.tsx [hunk 4]
```diff
@@ -239,30 +190,119 @@ function DataGrid<T>(props: {
...
+              anchorEl={filterMenuAnchorEl}
+              open={filterMenuAnchorEl !== null}
+              onClose={() => {
+                setFilterMenuAnchorEl(null)
+              }}
+            >
+              {filterChoices.map((choice) => (
+                <MenuItem
+                  key={choice}
+                  onClick={() => {
```
Comment (nabenabe0928) [inline optuna_dashboard/ts/components/DataGrid.tsx:268]: Is it better to separately define this function rather than define it here?

**Review by nabenabe0928 — COMMENTED** [review]
Thank you for your PR!
I made some minor comments")

Btw, shall we add the tests for filterChoices in another PR or shall we do it in this PR?

**Comment by nabenabe0928** [comment]
For the local test, I used the following:

```python
import optuna

import numpy as np


def create_study_with_various_trials(storage: str) -> None:
    rng = np.random.RandomState(42)

    def _objective_intermediate_values(trial: optuna.Trial) -> float:
        x = trial.suggest_categorical("x", choices=["a", "b", "c"])
        rnd = rng.random()
        if rnd < 0.25:
            trial.report(trial.number, step=0)
            trial.report(trial.number + 1, step=1)
        elif rnd < 0.5:
            trial.report(trial.number, step=0)
            raise optuna.TrialPruned()
        elif rnd < 0.75:
            raise optuna.TrialPruned()
        else:
            raise ValueError("Unexpected Error.")

        return 0.0

    study = optuna.create_study(study_name="various-trials", storage=storage)
    study.optimize(_objective_intermediate_values, n_trials=40, catch=(Exception,))

create_study_with_various_trials(storage=storage)
```

File: optuna_dashboard/ts/components/DataGrid.tsx [hunk 5]
```diff
@@ -81,49 +85,29 @@ function DataGrid<T>(props: {
...
-          if (toCellValue !== undefined) {
-            return toCellValue(rowIdx) === f.value
-          }
-          const field = columns[f.columnIdx].field
-          return row[field] === f.value
+          const value =
+            toCellValue !== undefined
+              ? toCellValue(rowIdx)
+              : row[columns[f.columnIdx].field]
+          return f.values.some((v) => v === value)
```
Comment (c-bata) [inline optuna_dashboard/ts/components/DataGrid.tsx:106]: To apply https://github.com/optuna/optuna-dashboard/pull/700/files#r1398672374

```suggestion
          const cellValue =
            toCellValue !== undefined
              ? toCellValue(rowIdx)
              : row[columns[f.columnIdx].field]
          return f.values.some((v) => v === cellValue)
```

File: optuna_dashboard/ts/components/DataGrid.tsx (hunk 4 above)
Comment (c-bata) [reply optuna_dashboard/ts/components/DataGrid.tsx:268]: I would say it's a matter of preference. Let me proceed without changes here.

File: optuna_dashboard/ts/components/DataGrid.tsx (hunk 3 above)
Comment (c-bata) [reply optuna_dashboard/ts/components/DataGrid.tsx:279]: Good catch! Thank you for pointing it out.

File: optuna_dashboard/ts/components/DataGrid.tsx [hunk 6]
```diff
@@ -239,30 +190,119 @@ function DataGrid<T>(props: {
...
+                <MenuItem
+                  key={choice}
+                  onClick={() => {
+                    const values =
+                      filter === null
+                        ? filterChoices.filter((v) => v !== choice)
+                        : filter.values.some((v) => v === choice)
+                        ? filter.values.filter((v) => v !== choice)
+                        : [...filter.values, choice]
+                    onFilterChange(values)
```
Comment (c-bata) [inline optuna_dashboard/ts/components/DataGrid.tsx:275]: To apply https://github.com/optuna/optuna-dashboard/pull/700/files#r1398699656.

```suggestion
                    const newTickedValues =
                      filter === null
                        ? filterChoices.filter((v) => v !== choice)
                        : filter.values.some((v) => v === choice)
                        ? filter.values.filter((v) => v !== choice)
                        : [...filter.values, choice]
                    onFilterChange(newTickedValues)
```

**Comment by c-bata** [comment]
@nabenabe0928 Thank you for your review. I applied your suggestions.

> Btw, shall we add the tests for filterChoices in another PR or shall we do it in this PR?

Yes, let us work on it in another PR.

**Review by nabenabe0928 — APPROVED** [review]
Thank you for the changes, LGTM!

**Pull request merged**